rfcs/20191017-tfx-standardized-inputs.md

```diff
@@ -600,68 +659,52 @@ be reused in a TF op (as the downstream of the Dataset, or in a Python wrapper).
-    our end-users (which might be the case eventually, but not likely to happen
-    soon). So a good portion of this proposal will remain mostly unchanged.
-*   We still cannot expose file-based or record-based interface to end-users in
-    the TF trainer. As that is a direct result of I/O + parsing being abstracted
-    out.
-
-The disadvantages of this approach are:
-
-*   This tightly couples TF with TFX components, in the following ways:
-
-    *   Components will need TF to read the data in.
-    *   Components that analyze the data (e.g. TFDV) will operate against
-        StructuredTensors, and the easiest way to to conduct certain
-        computations (for example, slicing a StructuredTensor or computing the
-        mean) with StructuredTensors is through TF ops and their python
-        bindings.
-
-    Such a coupling is not in our favor because:
-
-    *   Some TFX components functionally do not require TF to work. For example,
-        TFDV can analyze any data set. TFMA can analyze a model while treating
-        the model as a blackbox. In both cases, the ML library that trained the
-        model is irrelevant and TF should not be assumed. And coupling with TF
-        does not only introduce a heavy dependency, but also forces the user to
-        learn about TF if they need to implement a TFXIO for their data format.
-
-    *   Operations against StructuredTensors boils down to Python + TF ops. The
-        overhead of either is much higher than just calling an Arrow C++ API
-        that does the same operation.
-
-    *   The main extension point of TF is Ops which don't understand the nested
-        structures. Compared to using Arrow’s C++ APIs, implementing an Op that
-        deals with StructuredTensors will be much complicated.
+## StructuredTensor
+
+We’ve considered an alternative where [StructuredTensor](https://github.com/tensorflow/community/blob/master/rfcs/20190910-struct-tensor.md)
+is the unified in-memory representation, but it does not meet all of the
+requirements:
+
+|                                   | Arrow RecordBatch  | StructuredTensor  |
+| :-------------------------------- | :----------------- | :---------------- |
+| Columnar                          | Yes                | Yes               |
+| Lossless encoding (nullity)       | Yes                | No (see remark 1) |
+| Efficient translatable to Tensors | Yes (see remark 2) | Yes               |
+| Efficient slicing, filtering and concatenation | Yes                | No (see remark 3) |
+| Interoperability with the rest of the world | Good through Apache Arrow | Needs adaptation (see remark 4) |
+
+Remarks:
+
+1.  We could revise the design of StructuredTensor to include the nullibility
+    support.
+2.  Only when the backing buffers are aligned correctly. Currently both TF
+    and Apache Arrow has 64-byte alignment. And this can be enforced by
+    implementing our own Arrow MemoryPool wrapping a TF allocator.
```

Comment:
It would be useful to see a comparison here similar to the one in (3)

Reply:
I built another notebook to show that feeding an Arrow array (through the arrow->numpy bridge provided by Arrow) is not worse than feeding directly an aligned numpy array.
However it's a bit slower than just feeding an EagerTensor (which I would expect). But feeding an unaligned numpy array is >5x slower.

https://colab.sandbox.google.com/drive/1VKAQUPGB-bmaMJZUW5oi4JDwFt7Qirmh#scrollTo=bsHNuG_h6MD7